### PR TITLE
Multiple addresses per coin

### LINKED ISSUE #1788: [Keystore] Support multiple accounts per coin
**Describe the bug**

Currently our keystore always expects to only have 1 account/address per wallet. Bitcoin for example already supports 4 different types of addresses: 1,3, bc1 and taproot. It would be great to provide an interface to generate each account and then allow clients to select which address they want to use. 

Another use case to allow users to have multiple accounts with different derivation paths, Phantom wallet as an example, which using different derivation path. 

Clarification: one keystore/wallet still has one mnemonic / master private key.

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Prototype branch+PR for various ideas for multi-address support, not all changes will be necessarily included. See #1788 .

Changes:
- Multiple derivations can be defined for a coin, non-default is named.
- A derivation includes a derivation path.
- An address implementation can handle different address formats (per derivation).
- StoreKey supports multiple accounts per coin, of different derivation/address.

Breaking Changes:
- Method TWAccountCreate extended with a derivation parameter (method rarely used).

The following cases are used as test examples:
- Bitcoin has two address formats: default segwit, and p2pk.
- Solana has two derivation paths: default, and phantom wallet.

This appears in `registry.json` as follows:
```json
    "derivation": [
      {
        "name": "segwit",
        "path": "m/84'/0'/0'/0/0"
      },
      {
        "name": "legacy",
        "path": "m/44'/0'/0'/0/0"
      }
    ],
```

```json
    "derivation": [
      {
        "path": "m/44'/501'/0'"
      },
      {
        "name": "solana",
        "path": "m/44'/501'/0'/0'"
      }
    ],
```

Here is the example swift code:

```swift
    let btc1 = try wallet.getAccount(password: password, coin: .bitcoin, derivation: .default)
    XCTAssertEqual(btc1.address, "bc1qturc268v0f2srjh4r2zu4t6zk4gdutqd5a6zny")
    let btc2 = try wallet.getAccount(password: password, coin: .bitcoin, derivation: .bitcoinLegacy)
    XCTAssertEqual(btc2.address, "19fUCoUeGmHRFSgFtv4hoMYatCHcifNDEy")

    let solana1 = try wallet.getAccount(password: password, coin: .solana, derivation: .default)
    XCTAssertEqual(solana1.address, "HiipoCKL8hX2RVmJTz3vaLy34hS2zLhWWMkUWtw85TmZ")
    XCTAssertEqual(solana1.derivationPath, "m/44'/501'/0'")
    let solana2 = try wallet.getAccount(password: password, coin: .solana, derivation: .solanaSolana)
    XCTAssertEqual(solana2.address, "CgWJeEWkiYqosy1ba7a3wn9HAQuHyK48xs3LM4SSDc1C")
    XCTAssertEqual(solana2.derivationPath, "m/44'/501'/0'/0'")
```

## Checklist

- [x] Prefix PR title with `[WIP]` if necessary.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain
- [x] If there is a related Issue, mention it in the description (e.g. Fixes #<issue_number> ).
- [ ] Extend AnyAddress with derivation

### CODE REVIEW

File: src/Keystore/Account.cpp [hunk 1]
```diff
@@ -24,6 +25,10 @@ namespace CodingKeys {
 } // namespace CodingKeys
 
 Account::Account(const nlohmann::json& json) {
+    if (json.find(CodingKeys::derivation) != json.end()) {
+        derivation = TWDerivation(json[CodingKeys::derivation].get<uint32_t>());
+    }
+
     if (json[CodingKeys::derivationPath].is_object()) {
```
Comment (hewigovens) [inline src/Keystore/Account.cpp:32]:
```suggestion
    } else if (json[CodingKeys::derivationPath].is_object()) {
```

**Review by hewigovens — COMMENTED** [review]
Left some comments, and a few questions:

1. Could you please show that what will the new keystore json format will be like? 
2. Do we need to migrate old keystore?
3. don't forget to update `newcoin` script and `registry.json` description on developer.trustwallet.com

**Comment by catenocrypt** [comment]
> Left some comments, and a few questions:
> 
>     1. Could you please show that what will the new keystore json format will be like?
> 
>     2. Do we need to migrate old keystore?

For the keystore questions:
Keystore format has supported multiple accounts even before, so no real change there.
The change is backwards compatible regarding keystore.json files.
In case of non-default derivations, an additional field like `"derivation": 5` is added, but this is informational only.
The enum numerical values may change in the future, so no decisions should be made on this.
When a particular derivation+coin account is requested, the address is derived, and then the address+coin is used to select the right account (i.e., not the derivation value).
There is one exception: when account is retrieved for a coin, and there are multiple ones, we attempt to return the default.  If wallet is no available, and no address derivation is possible, only then the derivation==0 condition is used to return the default.  But this is a special case, and we may remove this extra logic.

**Comment by catenocrypt** [comment]
Here is the sample generated `TWDerivation.h` header file:

```cpp
// Copyright © 2017-2021 Trust Wallet.
//
// This file is part of Trust. The full Trust copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.
//
// This is a GENERATED FILE from \registry.json, changes made here WILL BE LOST.
//

#pragma once

#include "TWBase.h"

TW_EXTERN_C_BEGIN

/// Non-default coin address derivation names (default, unnamed derivations are not included).
TW_EXPORT_ENUM()
enum TWDerivation {
    TWDerivationDefault = 0, // default, for any coin
    TWDerivationCustom = 1, // custom, for any coin
    TWDerivationBitcoinSegwit = 2,
    TWDerivationBitcoinP2pk = 3,
    TWDerivationSolanaPhantom = 4,
};

TW_EXTERN_C_END
```

File: codegen/lib/templates/TWDerivation.h.erb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+#pragma once
+
+#include "TWBase.h"
+
+TW_EXTERN_C_BEGIN
+
+/// Non-default coin address derivation names (default, unnamed derivations are not included).
+TW_EXPORT_ENUM()
+enum TWDerivation {
```
[the remark this answers is not in the document]
Comment (catenocrypt) [reply codegen/lib/templates/TWDerivation.h.erb:18]: see below   https://github.com/trustwallet/wallet-core/pull/1816#issuecomment-993555512

File: src/Keystore/Account.cpp (hunk 1 above)
Comment (catenocrypt) [reply src/Keystore/Account.cpp:32]: Previous if is for derivation field, this is for derivationPath, should be independent.

**Review by costeazabica11213 — APPROVED** [review]

**Comment by costeazabica11213** [comment]
image 💰

File: registry.json [hunk 3]
```diff
@@ -6,7 +6,16 @@
...
     "blockchain": "Bitcoin",
-    "derivationPath": "m/84'/0'/0'/0/0",
+    "derivation": [
+      {
+        "name": "segwit",
+        "path": "m/84'/0'/0'/0/0"
+      },
+      {
+        "name": "p2pk",
+        "path": "m/84'/0'/0'/0/0"
```
[the remark this answers is not in the document]
Comment (catenocrypt) [reply registry.json]: Renamed.  Adding to Litecoin too, but not other coin for now.

File: src/interface/TWAccount.cpp [hunk 4]
```diff
@@ -15,7 +15,7 @@ struct TWAccount *_Nonnull TWAccountCreate(TWString *_Nonnull address, enum TWCo
     auto& derivationPathString = *reinterpret_cast<const std::string*>(derivationPath);
     auto& extendedPublicKeyString = *reinterpret_cast<const std::string*>(extendedPublicKey);
     const auto dp = DerivationPath(derivationPathString);
-    return new TWAccount{ Keystore::Account(addressString, coin, dp, extendedPublicKeyString) };
+    return new TWAccount{ Keystore::Account(addressString, coin, TWDerivationDefault, dp, extendedPublicKeyString) };  // TODO add version with alternative derivation
```
[the remark this answers is not in the document]
Comment (catenocrypt) [reply src/interface/TWAccount.cpp]: Handled in this PR

**Review by SSGoku369 — COMMENTED** [review]
Yes?

**Comment by catenocrypt** [comment]
Needs review

**Comment by hewigovens** [comment]
Please mark ready review otherwise there is no notification @catenocrypt

**Comment by catenocrypt** [comment]
Latest master merge made builds to fail, I'll check

**Comment by catenocrypt** [comment]
Auto-merge of master created a compile error, fixed.

**Review by madcake — APPROVED** [review]

**Comment by hewigovens** [comment]
@catenocrypt I will take over this pr, need to check iOS code and make sure it works as app expects

**Comment by hewigovens** [comment]
@catenocrypt wallet core side seems ready, it needs app to adopt this change (expects only one account / address for each coin, UX wise for Solana etc), could you think of a bit how app should adopt it?

**Comment by Princeobinna** [comment]
Unblock Merging

**Review by Princeobinna — APPROVED** [review]
Approve transactions

**Comment by hewigovens** [comment]
@catenocrypt have you exposed xpub for bip44 derivation?

**Review by ArtemGoryunov — APPROVED** [review]

**Review by hewigovens — APPROVED** [review]

**Review by Princeobinna — COMMENTED** [review]
Approve all assets

**Review by Princeobinna — COMMENTED** [review]
Approve assets

**Pull request merged**